modules/host/negotiateformcontract.go

```diff
@@ -130,14 +125,54 @@ func (h *Host) managedAddCollateral(txnSet []types.Transaction, settings modules
 // collateral, and then try submitting the file contract to the transaction
 // pool. If there is no error, the completed transaction set will be returned
 // to the caller.
-func (h *Host) managedFinalizeContract(builder modules.TransactionBuilder, renterSignatures []types.TransactionSignature) ([]types.TransactionSignature, error) {
+func (h *Host) managedFinalizeContract(builder modules.TransactionBuilder, renterPK crypto.PublicKey, renterSignatures []types.TransactionSignature, renterRevisionSignature types.TransactionSignature) ([]types.TransactionSignature, types.TransactionSignature, error) {
 	for _, sig := range renterSignatures {
 		builder.AddTransactionSignature(sig)
 	}
 	fullTxnSet, err := builder.Sign(true)
 	if err != nil {
 		builder.Drop()
-		return nil, err
+		return nil, types.TransactionSignature{}, err
+	}
+
+	// Verify that the signature for the revision from the renter is correct.
+	h.mu.RLock()
+	blockHeight := h.blockHeight
+	hostPK := h.publicKey.Key
+	hostSK := h.secretKey
+	h.mu.RUnlock()
+	contractTxn := fullTxnSet[len(fullTxnSet)-1]
+	fc := contractTxn.FileContracts[0]
+	noOpRevision := types.FileContractRevision{
+		ParentID: contractTxn.FileContractID(0),
+		UnlockConditions: types.UnlockConditions{
+			PublicKeys: []types.SiaPublicKey{
+				{
+					Algorithm: types.SignatureEd25519,
+					Key:       renterPK[:],
+				},
+				{
+					Algorithm: types.SignatureEd25519,
+					Key:       hostPK[:],
```

Comment:
presumably, the host will reject a file contract if the renter uses an unrecognized algorithm. So it can guarantee that it will always be able to create proofs.